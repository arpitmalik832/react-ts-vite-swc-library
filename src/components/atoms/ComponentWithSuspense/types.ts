export interface ComponentWithSuspenseProps {
  component: React.ReactNode;
  fallback?: React.ReactNode;
}
