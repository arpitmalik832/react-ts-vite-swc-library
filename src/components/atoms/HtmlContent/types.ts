export interface HtmlContentProps {
  title: string;
  description: string;
}
