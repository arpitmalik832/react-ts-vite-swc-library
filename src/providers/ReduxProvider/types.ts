import type { configureStore } from '@reduxjs/toolkit';

export interface ReduxProviderProps {
  children?: React.ReactNode;
  store: ReturnType<typeof configureStore>;
}
