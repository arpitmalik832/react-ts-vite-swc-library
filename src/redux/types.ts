import type { AxiosInstance } from 'axios';

import type {
  AllParams,
  KeyValuePair,
  VoidFunctionWithParams,
} from '../types/types';

export interface AppRedux extends Record<string, AllParams> {
  theme: string;
}

export interface APIData {
  host: string;
  headers: Record<string, string | Record<string, string>>;
  axiosInstance: AxiosInstance;
}

export interface UpdateApiHostByValue {
  oldValue: string;
  newValue: string;
}

export interface UpdateApiHostByIndex {
  index: number;
  newValue: string;
}

export interface UpdateApiHeadersByHost {
  host: string;
  newHeaders: Record<string, string | Record<string, string>>;
}

export interface UpdateApiHeadersByIndex {
  index: number;
  newHeaders: Record<string, string | Record<string, string>>;
}

export interface AddToApiHeadersByHost {
  host: string;
  newHeader: KeyValuePair<string>;
}

export interface AddToApiHeadersByIndex {
  index: number;
  newHeader: KeyValuePair<string>;
}

export interface UpdateApiAxiosInstanceByHost {
  host: string;
  axiosInstance: AxiosInstance;
}

export interface UpdateApiAxiosInstanceByIndex {
  index: number;
  axiosInstance: AxiosInstance;
}

export type ApisRedux = APIData[];

export interface NavigationRedux {
  stack: VoidFunctionWithParams[];
}

export interface ReduxState {
  app: AppRedux;
  apis: ApisRedux;
  navigation: NavigationRedux;
}

export interface BaseQueryParams {
  axiosInstance: AxiosInstance;
  url: string;
}
