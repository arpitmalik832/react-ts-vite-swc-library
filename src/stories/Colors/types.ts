export type ColorTheme = 'light' | 'dark';
export type ColorType = 'background';
export type ColorInnerType = 'primary';
export type PrimitiveColor = 'black' | 'white';
export type ColorSemanticLabel = '50' | '100' | '900';
export type HexColor = `#${string & { length: 6 | 8 }}`;

export interface DesignTokens {
  'color-primitives': Record<
    PrimitiveColor,
    Record<
      ColorSemanticLabel,
      {
        type: 'value';
        color: HexColor;
      }
    >
  >;
  'color-semantics': Record<
    ColorTheme,
    Record<
      ColorType,
      Record<
        ColorInnerType,
        Record<
          ColorSemanticLabel,
          {
            type: 'value';
            color: HexColor;
          }
        >
      >
    >
  >;
}
