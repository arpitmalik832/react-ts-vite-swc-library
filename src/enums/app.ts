const ENVS = {
  DEV: 'development',
  STG: 'staging',
  BETA: 'beta',
  PROD: 'production',
};

const THEME = {
  LIGHT: 'light',
  DARK: 'dark',
};

const API1_TIMEOUT = 15000;

export { ENVS, THEME, API1_TIMEOUT };
