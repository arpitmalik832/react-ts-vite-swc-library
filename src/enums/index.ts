export * as APP from './app';
export * as SW from './sw';
