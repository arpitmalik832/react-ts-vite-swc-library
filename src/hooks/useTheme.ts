import { useCallback, useLayoutEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { setDarkTheme, setLightTheme } from '../redux/slices/appSlice';
import preferredColorScheme from '../utils/eventListeners/preferredColorScheme';
import { THEME } from '../enums/app';
import { ReduxState, AppRedux } from '../types/types.d';

function useTheme() {
  const { theme } = useSelector<ReduxState, AppRedux>(state => state.app);
  const dispatch = useDispatch();

  const updateStore = useCallback((isDark: boolean) => {
    if (isDark) {
      if (theme !== THEME.DARK) {
        dispatch(setDarkTheme(undefined));
      }
    } else if (theme !== THEME.LIGHT) {
      dispatch(setLightTheme(undefined));
    }
  }, []);

  useLayoutEffect(() => {
    preferredColorScheme.subscribe(e => {
      updateStore(e.matches);
    });

    return () => {
      preferredColorScheme.unSubscribe();
    };
  }, []);
}

export default useTheme;
