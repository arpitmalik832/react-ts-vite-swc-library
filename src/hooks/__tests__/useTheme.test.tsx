import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  configureStore,
  createSlice,
  SliceCaseReducers,
  SliceSelectors,
} from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import preferredColorScheme from '../../utils/eventListeners/preferredColorScheme';
import { THEME } from '../../enums/app';
import useTheme from '../useTheme';
import { AppRedux } from '../../types/types.d';

jest.mock('../../utils/eventListeners/preferredColorScheme', () => ({
  __esModule: true,
  default: {
    subscribe: jest.fn(),
    unSubscribe: jest.fn(),
  },
}));

describe('useAppMount unit tests', () => {
  afterEach(cleanup);

  it('incase of light mode', () => {
    (preferredColorScheme.subscribe as jest.Mock).mockImplementation(
      callBackFn => callBackFn({ matches: false }),
    );

    const appSlice = createSlice<
      AppRedux,
      SliceCaseReducers<AppRedux>,
      string,
      SliceSelectors<AppRedux>,
      string
    >({
      name: 'app',
      initialState: {
        theme: THEME.LIGHT,
      },
      reducers: {},
    });

    const store = configureStore({
      reducer: {
        app: appSlice.reducer,
      },
    });

    function TestComponent() {
      useTheme();

      return <div data-testid="temp-component">Test Component</div>;
    }

    const component = render(
      <Provider store={store}>
        <TestComponent />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });

  it('incase of theme changes from dark mode to light mode', () => {
    (preferredColorScheme.subscribe as jest.Mock).mockImplementation(
      callBackFn => callBackFn({ matches: true }),
    );

    const appSlice = createSlice<
      AppRedux,
      SliceCaseReducers<AppRedux>,
      string,
      SliceSelectors<AppRedux>,
      string
    >({
      name: 'app',
      initialState: {
        theme: THEME.DARK,
      },
      reducers: {},
    });

    const store = configureStore({
      reducer: {
        app: appSlice.reducer,
      },
    });

    function TestComponent() {
      useTheme();

      return <div data-testid="temp-component">Test Component</div>;
    }

    const component = render(
      <Provider store={store}>
        <TestComponent />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });

  it('incase of theme changes from light mode to dark mode', () => {
    (preferredColorScheme.subscribe as jest.Mock).mockImplementation(
      callBackFn => callBackFn({ matches: true }),
    );

    const appSlice = createSlice<
      AppRedux,
      SliceCaseReducers<AppRedux>,
      string,
      SliceSelectors<AppRedux>,
      string
    >({
      name: 'app',
      initialState: {
        theme: THEME.LIGHT,
      },
      reducers: {},
    });

    const store = configureStore({
      reducer: {
        app: appSlice.reducer,
      },
    });

    function TestComponent() {
      useTheme();

      return <div data-testid="temp-component">Test Component</div>;
    }

    const component = render(
      <Provider store={store}>
        <TestComponent />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });

  it('incase of dark mode', () => {
    (preferredColorScheme.subscribe as jest.Mock).mockImplementation(
      callBackFn => callBackFn({ matches: false }),
    );

    const appSlice = createSlice<
      AppRedux,
      SliceCaseReducers<AppRedux>,
      string,
      SliceSelectors<AppRedux>,
      string
    >({
      name: 'app',
      initialState: {
        theme: THEME.DARK,
      },
      reducers: {},
    });

    const store = configureStore({
      reducer: {
        app: appSlice.reducer,
      },
    });

    function TestComponent() {
      useTheme();

      return <div data-testid="temp-component">Test Component</div>;
    }

    const component = render(
      <Provider store={store}>
        <TestComponent />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });
});
