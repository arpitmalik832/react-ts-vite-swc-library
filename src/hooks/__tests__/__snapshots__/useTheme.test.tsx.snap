// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAppMount unit tests snapshot test 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": undefined,
  },
  "unmount": [Function],
}
`;
